crates/transaction-pool/src/validate.rs

```diff
@@ -80,6 +77,137 @@ pub trait TransactionValidator: Send + Sync {
     }
 }
 
+pub(crate) struct EthTransactionValidatorConfig<Client: AccountProvider> {
+    /// Chain id
+    chain_id: Option<u64>,
+    /// This type fetches account info from the db
+    client: Client,
+    /// Fork indicator whether we are in the Shanghai stage.
+    shanghai: bool,
+    /// Fork indicator whether we are using EIP-2718 type transactions.
+    eip2718: bool,
+    /// Fork indicator whether we are using EIP-1559 type transactions.
+    eip1559: bool,
+    /// The current max gas limit
+    current_max_gas_limit: u64,
+    /// gasprice
+    gas_price: Option<u128>,
+}
+
+#[async_trait::async_trait]
+impl<T: PoolTransaction + AccountProvider> TransactionValidator
+    for EthTransactionValidatorConfig<T>
+{
+    type Transaction = T;
+
+    async fn validate_transaction(
+        &self,
+        origin: TransactionOrigin,
+        transaction: Self::Transaction,
+    ) -> Result<TransactionValidationOutcome<Self::Transaction>, Error> {
+        let invalid_transaction = transaction.clone();
+
+        match transaction.tx_type() {
+            0 => {
+                // Accept legacy transactions
+            }
+
+            1 => {
+                // Accept only legacy transactions until EIP-2718/2930 activates
+                if !self.eip2718 {
+                    return Err(Error::TransactionEip2930Disabled)
+                }
+            }
+
+            2 => {
+                // Reject dynamic fee transactions until EIP-1559 activates.
+                if !self.eip1559 {
+                    return Err(Error::TransactionEip1559Disabled)
+                }
+            }
+
+            _ => return Err(Error::TxTypeNotSupported),
+        };
+
+        if self.shanghai {
+            match self.ensure_max_init_code_size(transaction.clone(), 2 * 24576) {
+                Ok(_) => {}
+                Err(e) => return Ok(TransactionValidationOutcome::Invalid(transaction, e)),
+            }
+        }
+
+        // Drop non-local transactions under our own minimal accepted gas price or tip
+        if !origin.is_local() && transaction.max_fee_per_gas() < self.gas_price {
+            return Err(Error::TransactionMaxFeeLessThenBaseFee)
+        }
+
+        // Ensure gasFeeCap is greater than or equal to gasTipCap.
+        if transaction.max_fee_per_gas() >= transaction.max_priority_fee_per_gas() {
+            return Err(Error::TransactionMaxFeeLessThenBaseFee)
+        }
```

Comment:
this error does not match the check.

likely needs new error variants, see also `InvalidTransactionError` which we should move perhaps

Reply:
> likely needs new error variants, see also `InvalidTransactionError` which we should move perhaps

Or we could use this error instead of consensus error?